File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    
    int solve(vector<int>& a,int n,vector<int>& dp)
    {
        if(n<0)return 0;
        
        if(n==0) return a[0];
        
        if(dp[n]!=-1)return dp[n];
        
        int include = solve(a,n-2,dp)+a[n];
        int exclude = solve(a,n-1,dp)+0;
        
        return dp[n]=max(include,exclude);
    }
    
    int rob(vector<int>& a) {
        vector<int>dp(a.size()+1,-1);
        return solve(a,a.size()-1,dp);
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        @lru_cache
        def rec(i):
            if i==0:
                return nums[i]
            if i<0:
                return 0
            pick = nums[i] + rec(i-2)
            not_pick = 0 + rec(i-1)
            return max(pick,not_pick)
        
        return rec(len(nums)-1)

    #sumitted by Vibhor chutiya